doc/sphinx/language/core/records.rst

```diff
@@ -3,208 +3,311 @@
-fields. For instance, we might have:
-:n:`Record @ident {* @binder } : @sort := { @ident__1 : @type__1 ; @ident__2 := @term__2 ; @ident__3 : @type__3 }`.
-in which case the correctness of :n:`@type__3` may rely on the instance :n:`@term__2` of :n:`@ident__2` and :n:`@term__2` may in turn depend on :n:`@ident__1`.
+   .. example:: Defining a record
+
+      The set of rational numbers may be defined as:
 
-.. example::
+      .. coqtop:: reset all
 
-   The set of rational numbers may be defined as:
+         Record Rat : Set := mkRat
+          { negative : bool
+          ; top : nat
+          ; bottom : nat
+          ; Rat_bottom_nonzero : 0 <> bottom
+          ; Rat_irreducible :
+              forall x y z:nat, (x * y) = top /\ (x * z) = bottom -> x = 1
+          }.
 
-   .. coqtop:: reset all
+      The :n:`Rat_*` fields depend on :n:`top` and :n:`bottom`.
+      :n:`Rat_bottom_nonzero` is a proof that :n:`bottom` (the denominator)
+      is not zero.  :n:`Rat_irreducible` is a proof that the fraction is in
+      lowest terms.
 
-      Record Rat : Set := mkRat
-       { sign : bool
-       ; top : nat
-       ; bottom : nat
-       ; Rat_bottom_cond : 0 <> bottom
-       ; Rat_irred_cond :
-           forall x y z:nat, (x * y) = top /\ (x * z) = bottom -> x = 1
-       }.
+.. _reuse_field_name:
 
-   Note here that the fields ``Rat_bottom_cond`` depends on the field ``bottom``
-   and ``Rat_irred_cond`` depends on both ``top`` and ``bottom``.
+   .. example:: Reusing a field name in multiple records
 
-Let us now see the work done by the :cmd:`Record` command. First the command
-generates a variant type definition with just one constructor:
-:n:`Variant @ident {* @binder } : @sort := @ident__0 {* @binder }`.
+      .. coqtop:: in
 
-To build an object of type :token:`ident`, provide the constructor
-:n:`@ident__0` with the appropriate number of terms filling the fields of the record.
+         Module A. Record R := { f : nat }. End A.
+         Module B. Record S := { f : nat }. End B.
 
-.. example::
+      .. coqtop:: all
 
-   Let us define the rational :math:`1/2`:
+         Check {| A.f := 0 |}.
+         Check {| B.f := 0 |}.
 
-    .. coqtop:: in
+.. _record_as_clause:
 
-       Theorem one_two_irred : forall x y z:nat, x * y = 1 /\ x * z = 2 -> x = 1.
-       Admitted.
+   .. example:: Using the "as" clause in a record definition
 
-       Definition half := mkRat true 1 2 (O_S 1) one_two_irred.
-       Check half.
+      .. coqtop:: all
 
-Alternatively, the following syntax allows creating objects by using named fields, as
-shown in this grammar. The fields do not have to be in any particular order, nor do they have
-to be all present if the missing ones can be inferred or prompted for
-(see :ref:`programs`).
+         Record foo := { bar : nat } as HIMOM.
+         About bar.
+         Arguments bar {HIMOM}.   (* make "foo" an implicit parameter *)
+         Check bar.
+         Check (bar (HIMOM:={|bar:=0|})).
```

Comment:
The names here aren't great, though they were sufficient for discussion. How about:

```suggestion
         Record MyRecord := { myfield : nat } as VarName.
         About myfield. (* observe the MyRecord variable is named VarName *)
         
         (* This allows us to make VarName implicit for instance, without having to rename the variable as would be the case without the as clause. *)
         Arguments myfield {VarName}.   (* make "VarName" an implicit parameter *)
         Check myfield.
         Check (myfield (VarName:={| myfield := 0 |})).
```